codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +444,125 @@ public Http2Settings pollSentSettings() {
     public Configuration configuration() {
         return frameWriter.configuration();
     }
+
+    /**
+     * Wrap a DATA frame so it can be written subject to flow-control. Note that this implementation assumes it
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledPayloadBase {
+        private final ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, padding, endOfStream, streamId, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
```

Comment:
> ~~Thinking about it more I'm not sure my suggestion makes sense because if we are passing in `data` directly then it will be released and we wouldn't want to use any methods on it.~~

Ignore this.  We can just call `retain()`.  That way we can get rid of `size` and the independent tracking required to maintain it.